### ISSUE TITLE
Can't search for empty genre

### ISSUE THREAD

**Issue by Clementine-Issue-Importer** [clementine-player/Clementine#2617]
_From [alphadel...@gmail.com](https://code.google.com/u/107136459110240960141/) on January 16, 2012 09:08:52_

What steps will reproduce the problem? 1.  Make a new smart playlist
2.  Add a search term for "Genre equals" and leave the third field empty What is the expected output? What do you see instead? I expect to see only tracks which have no set Genre.  Instead I see all tracks regardless of their Genre field. What version of the product are you using? On what operating system? 1.0 on Linux Please provide any additional information below. This app is the best music player, hands down.  Amarok 1.4 is dead!  Long live Clementine!

_Original issue: http://code.google.com/p/clementine-player/issues/detail?id=2617_

**Comment by kleinfreund** [comment]
Smart Playlists currently can’t contain empty values in the Search Terms field.

It would be useful to either

- be able to leave the search term empty or
- have `is empty` and `is not empty` qualifiers

This would allow users to filter out tracks that are part of an album or tracks that don’t have a genre, yet.

Any chance this will get implemented?

**Comment by kleinfreund** [comment]
I cloned the repository and tried implementing this on my own.

My first naive approach was changing the following line to `return true;`: https://github.com/clementine-player/Clementine/blob/e5ab3e786f9adde12cec3cc90cfe8c1cc6b06320/src/smartplaylists/searchterm.cpp#L146

This will crash the program when opening the create smart playlist dialog.

Next, I tried to add the operators `Op_Empty` and `Op_NotEmpty`. However, I didn’t find an easy way to remove the text field for these operators. Currently, the type of the current search term is determined by the field (e.g. Artist, [see here](https://github.com/clementine-player/Clementine/blob/e5ab3e786f9adde12cec3cc90cfe8c1cc6b06320/src/smartplaylists/searchterm.cpp#L167-L196)). The default type is `Type_Text`; however, `Op_Empty` and `Op_NotEmpty` are not of type text. Yet, they need to appear in the same combobox along with operators that require a text field.

**TL;DR:** The type of the input field should be determined by the selected operator, not the selected field. This would allow mixing input field types.

**Comment by kleinfreund** [comment]
This was resolved in #5985.
